### PR TITLE
Add a Book Builder class

### PR DESCRIPTION
**Pull request by bghughes**
The goal of this class is to provide a ready-to-use object that **makes it easy to track open orders for any given ERC20/ERC20 Gladius book**

Two features not yet added, that should be added:
- [x] Lead-up time. An additional deadline filter where orders that are Date.now() + LeadUpTime away are not considered. This is a consideration for users who may need to, for instance, offer aggregator quotes to users that must be valid for LeadUpTime seconds into the feature. Should default to 0.
- [x] Market Buy and Market Sell helper functions. These functions should return the exact data required to execute a market buy or sell into any given book the BookBuilder is watching. Importantly, these methods must adhere to the **no partial fills** condition of Gladius

### CODE REVIEW

**Comment by bghughes** [comment]
Another feature for future improvement:
- [ ]  Add a feature that periodically uses `multicall` to batch check outstanding orders in the BookBuilder and automatically remove those that may have become in-`valid` in normal course of action

**Review by bghughes — COMMENTED** [review]
Ready to go from my eyes

**Review by etherealiska — CHANGES_REQUESTED** [review]

File: src/utils/BookBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            }
+        }, this.gladiusPollTime);
+    }
+
+    /**
+    * Fetches offers from the Gladius service for both bids and asks.
+    * @public
+    * @returns {Promise<{bids: any; asks: any}>} An object containing bid and ask data in raw form, unparsed. See parseOrderWithChainID or parseOrder for parsing.
+    */
+    public async getOffers(): Promise<{ bids: any; asks: any }> {
```
[the remark this answers is not in the document]
Comment (bghughes) [reply src/utils/BookBuilder.ts:126]: There isn't a type already for this one... you're totally right though...

[the remark this answers is not in the document]
Comment (bghughes) [reply src/utils/BookBuilder.ts:126]: worth noting it is only used internally 😅

[the remark this answers is not in the document]
Comment (bghughes) [reply src/utils/BookBuilder.ts:126]: like most users will just `getBook`

**Pull request merged**